fbx/FBXSemantic.g4

```diff
@@ -0,0 +1,588 @@
+      CLOSE_BRACE
+    ;
+
+nodeP
+    : 'P'
+      COLON propertyName=STRING
+      COMMA propertyType1=STRING
+      COMMA propertyType2=STRING
+      COMMA propertyValue=anyValueList
+    ;
+
+nodeConnections
+    : 'Connections' COLON OPEN_BRACE
+      ( nodeConnect | nodeC )*
+      CLOSE_BRACE
+    ;
+
+nodeConnect: 'Connect' COLON x=STRING COMMA from=STRING COMMA to=STRING;
+
+nodeC: 'C' COLON x=STRING COMMA from=INTEGER COMMA to=INTEGER ( COMMA connectType=STRING )?;
+
+nodeTakes
+    : 'Takes' COLON OPEN_BRACE
+      nodeCurrent?
+      nodeTake*
+      CLOSE_BRACE
+    ;
+
+nodeTake
+    : 'Take' COLON name=STRING OPEN_BRACE
+      nodeFileName
+      nodeLocalTime
+      nodeReferenceTime
+      CLOSE_BRACE
+    ;
+
+nodeFileName: 'FileName' COLON value=STRING;
+nodeLocalTime: 'LocalTime' COLON value=integerList;
+nodeReferenceTime: 'ReferenceTime' COLON value=integerList;
+
+nodeCurrent: 'Current' COLON value=STRING;
+
+nodeVersion5
+    : 'Version5' COLON OPEN_BRACE
+      nodeAmbientRenderSettings
+      nodeFogOptions
+      nodeSettings
+      nodeRendererSetting
+      CLOSE_BRACE
+    ;
+
+nodeAmbientRenderSettings
+    : 'AmbientRenderSettings' COLON OPEN_BRACE
+      nodeVersion
+      nodeAmbientLightColor
+      CLOSE_BRACE
+    ;
+
+nodeAmbientLightColor: 'AmbientLightColor' COLON value=decimalList;
+
+nodeFogOptions
+    : 'FogOptions' COLON OPEN_BRACE
+      nodeFogEnable
+      nodeFogMode
+      nodeFogDensity
+      nodeFogStart
+      nodeFogEnd
+      nodeFogColor
+      CLOSE_BRACE
+    ;
+
+nodeFogEnable: 'FogEnable' COLON value=INTEGER;
+nodeFogMode: 'FogMode' COLON value=INTEGER;
+nodeFogDensity: 'FogDensity' COLON value=DECIMAL;
+nodeFogStart: 'FogStart' COLON value=DECIMAL;
+nodeFogEnd: 'FogEnd' COLON value=DECIMAL;
+nodeFogColor: 'FogColor' COLON value=decimalList;
+
+nodeSettings
+    : 'Settings' COLON OPEN_BRACE
+      nodeFrameRate
+      nodeTimeFormat
+      nodeSnapOnFrames
+      nodeReferenceTimeIndex
+      nodeTimeLineStartTime
+      nodeTimeLineStopTime
+      CLOSE_BRACE
+    ;
+
+nodeFrameRate: 'FrameRate' COLON value=STRING;
+nodeTimeFormat: 'TimeFormat' COLON value=INTEGER;
+nodeSnapOnFrames: 'SnapOnFrames' COLON value=INTEGER;
+nodeReferenceTimeIndex: 'ReferenceTimeIndex' COLON value=INTEGER;
+nodeTimeLineStartTime: 'TimeLineStartTime' COLON value=INTEGER;
+nodeTimeLineStopTime: 'TimeLineStopTime' COLON value=INTEGER;
+
+nodeRendererSetting
+    : 'RendererSetting' COLON OPEN_BRACE
+      nodeDefaultCamera
+      nodeDefaultViewingMode
+      CLOSE_BRACE
+    ;
+
+nodeDefaultCamera: 'DefaultCamera' COLON value=STRING;
+nodeDefaultViewingMode: 'DefaultViewingMode' COLON value=INTEGER;
+
+nodeVersion: 'Version' COLON value=INTEGER;
+
+integerList: ( integer ( COMMA integer )* )?;
+integer: value=INTEGER;
+
+decimalList: ( decimal ( COMMA decimal )* )?;
+decimal: value=( INTEGER | DECIMAL );
+
+anyValueList: ( anyValue ( COMMA anyValue )* )?;
+anyValue: value=( INTEGER | DECIMAL | STRING );
```

Comment:
Also, remove optinality in declarations and use optionality in use-sites.

Reply:
I don't think this would help. My reasoning:

1. After I move the `?` to the use site, I will then have repetition because 100% of the places where you can insert an integer list, it can be completely empty.
2. So I would then want to create an integerListOrEmpty to reduce the duplication. 100% of the places would then use that instead.
3. Seeing that there is only one remaining usage for integerList, I would then inline that.
4. The end result is - the same rules, but now "integerList" is called "integerListOrEmpty" - I rename it for brevity and we're back where we started. :(